packages/typespec-ts/src/modular/serialization/buildSerializerFunction.ts

```diff
@@ -331,37 +343,90 @@ function buildModelTypeSerializer(
-    ? "...item"
-    : "";
+  if (
+    (type.usage & UsageFlags.Input) === UsageFlags.Input &&
+    (type.usage & UsageFlags.MultipartFormData) === UsageFlags.MultipartFormData
+  ) {
+    // For MFD models, serialize into an array of parts
+    // TODO: cleaner abstraction, quite a bit of duplication with the non-MFD stuff here
+    const parts: string[] = [];
 
-  const { directAssignment, propertiesStr } = getRequestModelMapping(
-    context,
-    type,
-    "item"
-  );
-  if (additionalPropertiesSpread) {
-    propertiesStr.unshift(additionalPropertiesSpread);
-  }
-  const serializeContent =
-    directAssignment === true
-      ? propertiesStr.join(",")
-      : `{${propertiesStr.join(",")}}`;
+    const properties = getAllProperties(type, getAllAncestors(type));
+    for (const property of properties) {
+      if (property.kind !== "property") {
+        continue;
+      }
+      const expr = getSerializationExpression(context, property, "item");
+
+      let partDefinition: string;
+      if (property.isMultipartFileInput) {
+        const createFilePartDescriptorDefinition = resolveReference(
+          MultipartHelpers.createFilePartDescriptor
+        );
+
+        const itemPath = property.multipartOptions?.isMulti
+          ? "x"
+          : getPropertyFullName(context, property, "item");
+        partDefinition = `${createFilePartDescriptorDefinition}("${property.serializedName}", ${itemPath}, )`;
+
+        // If the TypeSpec doesn't specify a default content type, TCGC will infer a default of "*/*".
+        // In this case, we actually want the content type to be left unset so that Core will take care of
+        // setting the content type correctly.
+        const contentType =
+          property.multipartOptions?.defaultContentTypes?.[0] === "*/*"
+            ? undefined
+            : property.multipartOptions?.defaultContentTypes?.[0];
 
-  const output = [];
+        if (property.multipartOptions?.isMulti) {
+          partDefinition = `...(item["${property.serializedName}"].map((x: unknown) => ${createFilePartDescriptorDefinition}("${property.serializedName}", x${contentType ? `", ${contentType}"` : ""})))`;
+        } else {
+          partDefinition = `${createFilePartDescriptorDefinition}("${property.serializedName}", item["${property.serializedName}"]${contentType ? `, "${contentType}"` : ""})`;
+        }
+      } else if (property.multipartOptions?.isMulti) {
+        partDefinition = `...((${expr}).map((x: unknown) => ({ name: "${property.serializedName}", body: x })))`;
+      } else {
+        partDefinition = `{ name: "${property.serializedName}", body: (${expr}) }`;
+      }
+
+      if (property.optional) {
+        parts.push(
+          `...(${getPropertyFullName(context, property, "item")} === undefined ? [] : [${partDefinition}])`
+        );
+      } else {
+        parts.push(partDefinition);
+      }
+
+      // TODO: How to handle additionalProperties for MFD?
```

Comment:
is there an issue to track this?

Reply:
Added it to the follow up issue https://github.com/Azure/autorest.typescript/issues/3001